torch/csrc/jit/register_prim_ops.cpp

```diff
@@ -111,6 +112,26 @@ static at::Tensor to_dispatch(
 
 RegisterOperators reg(
     {Operator(
+         "prim::profile(Tensor t) -> ()",
+         [](const Node* node) {
+           return [node](Stack& stack) {
+             auto p_pr =
+                 reinterpret_cast<ProfilingRecord*>(node->i(attr::storage));
+             p_pr->record(node, stack);
+             return 0;
+           };
+         }),
+     Operator(
+         "prim::finish_profile() -> ()",
+         [](const Node* node) {
+           return [node](Stack& stack) {
+             auto p_pr =
+                 reinterpret_cast<ProfilingRecord*>(node->i(attr::storage));
+             p_pr->finish(node, stack);
+             return 0;
+           };
+         }),
+     Operator(
```

Comment:
I still think this might be cleaner if the only registered op was:

```
	     Operator(
         "prim::observe(...) -> ()",
         [](const Node* node) {
           size_t N = node->inputs().size();
           std::function<void(at::ArrayRef<IValue>)>* callback = reinterpret_cast<...>(node->i(attr::callback));
           return [node](Stack& stack) {
                (*callback)(last(stack, N));
             return 0;
           };
         }),
```
It would make it easier to change what the profiler does with observations in the future since the linking between the observed value and where to record it stays inside record profile. It would be  harder to that with the current implementation because it is unsafe to capture `node` as described above. I also am not sure that all observers really can fit as subclasses to ProfilingRecord. For instance, not all profilers would be recording information into a Graph, but ProfilingRecord always has a graph.

Reply:
fixed